src/Xamarin.Android.Build.Tasks/Tasks/CreateMultiDexMainDexClassList.cs

```diff
@@ -35,45 +42,70 @@ public override bool Execute ()
 			Log.LogDebugTaskItems ("  CustomMainDexListFiles:", CustomMainDexListFiles);
 			Log.LogDebugMessage ("  ToolExe: {0}", ToolExe);
 			Log.LogDebugMessage ("  ToolPath: {0}", ToolPath);
-			Log.LogDebugMessage ("  ProguardHome: {0}", ProguardHome);
+			Log.LogDebugMessage ("  ProguardJarPath: {0}", ProguardJarPath);
 
 			if (CustomMainDexListFiles != null && CustomMainDexListFiles.Any ()) {
 				var content = string.Concat (CustomMainDexListFiles.Select (i => File.ReadAllText (i.ItemSpec)));
 				File.WriteAllText (MultiDexMainDexListFile, content);
 				return true;
 			}
 
-			EnvironmentVariables = MonoAndroidHelper.GetProguardEnvironmentVaribles (ProguardHome);
+			tempJar = Path.Combine (Path.GetTempPath (), Path.GetRandomFileName () + ".jar");
+			commandlineAction = GenerateProguardCommands;
+			// run proguard first
+			var retval = base.Execute ();
+			if (!retval || Log.HasLoggedErrors)
+				return false;
+
+			commandlineAction = GenerateMainDexListBuilderCommands;
+			// run java second
+
+			return base.Execute () && !Log.HasLoggedErrors; 
 
-			return base.Execute ();
 		}
 
 		protected override string GenerateCommandLineCommands ()
 		{
 			var cmd = new CommandLineBuilder ();
+			commandlineAction (cmd);
+			return cmd.ToString ();
+		}
 
-			cmd.AppendSwitch ("--output");
-			cmd.AppendFileNameIfNotNull (MultiDexMainDexListFile);
-
+		void GenerateProguardCommands (CommandLineBuilder cmd)
+		{
+			var enclosingChar = OS.IsWindows ? "\"" : string.Empty;
 			var jars = JavaLibraries.Select (i => i.ItemSpec).Concat (new string [] { ClassesOutputDirectory });
-			string files = string.Join (Path.PathSeparator.ToString (), jars.Select (s => '\'' + s + '\''));
-			if (OS.IsWindows)
-				cmd.AppendSwitch ('"' + files + '"');
-			else
-				cmd.AppendSwitch (files);
-
-			return cmd.ToString ();
+			cmd.AppendSwitchIfNotNull ("-jar ", ProguardJarPath);
+			foreach (var jar in jars)
+				cmd.AppendSwitchIfNotNull ("-injars ", jar);
+			cmd.AppendSwitch ("-dontwarn");
+			cmd.AppendSwitch ("-forceprocessing");
+			cmd.AppendSwitchIfNotNull ("-outjars ", tempJar);
+			cmd.AppendSwitchIfNotNull ("-libraryjars ", $"'{Path.Combine (AndroidSdkBuildToolsPath, "lib", "shrinkedAndroid.jar")}'");
```

Comment:
Indeed, *totally* weird. This would make sense if we were going through some "intermediary," such as a batch file/shell script/"wrapper program" which "forwarded" command-line parameters w/o further quoting, but that shouldn't be the case here; we're invoking `java.exe`, aren't we?

:-/